woocommerce/woocommerce
# Cart Data Store: Prevent unnecessary requests to cart API when cart is empty (for the mini cart block)

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This is an updated approach to prevent API requests on load when using the mini-cart block. This supersedes #51981 

1. Removes the mini cart specific caching (local storage) and update logic
2. Makes the mini cart use the cart data store
3. Adds persistent caching to cart data store
4. Resolves early without an API request if the cart is empty (no cookie) or the local storage contains the current cart which matches the current cookie. Otherwise fires an API request as normal.

This should mitigate page caching issues with stale data without the extra request to the server when its not needed.

Closes #50793
Closes #51981



### How to test the changes in this Pull Request:

1. Add the Mini Cart Block to a page.
2. Ensure it renders the correct quantity in cart on page load. Add some items to the cart to test this.
3. With a new session and empty cart, view the page with the browser network inspector open. Ensure no XHR requests are made.
4. Add `?add-to-cart=16` (use a valid product ID) to the URL. Ensure mini cart is updated.
5. Open the shop page on another tab and add something to the cart. Refresh mini cart. It updates.



### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
